libs/utils/src/shared_retryable.rs

```diff
@@ -0,0 +1,647 @@
+///     )
+///       .await
+///   }
+///
+///   async fn work_once() -> Result<u8, OneLevelError> {
+///     Ok(42)
+///   }
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///   let svc = Service::default();
+///
+///   let mut js = tokio::task::JoinSet::new();
+///
+///   let barrier = Arc::new(tokio::sync::Barrier::new(10 + 1));
+///   let completions = Arc::new(std::sync::atomic::AtomicUsize::new(0));
+///
+///   let handles = (0..10).map(|_| js.spawn({
+///     let svc = svc.clone();
+///     let barrier = barrier.clone();
+///     let completions = completions.clone();
+///     async move {
+///       // make sure all tasks are ready to start at the same time
+///       barrier.wait().await;
+///       // after successfully starting the work, any of the futures could get cancelled
+///       svc.work(completions).await
+///     }
+///   })).collect::<Vec<_>>();
+///
+///   barrier.wait().await;
+///
+///   tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+///
+///   handles[5].abort();
+///
+///   let mut cancellations = 0;
+///
+///   while let Some(res) = js.join_next().await {
+///     // all complete with the same result
+///     match res {
+///       Ok(res) => assert_eq!(res.unwrap(), 42),
+///       Err(je) => {
+///         // except for the one task we cancelled; it's cancelling
+///         // does not interfere with the result
+///         assert!(je.is_cancelled());
+///         cancellations += 1;
+///         assert_eq!(cancellations, 1, "only 6th task was aborted");
+///       }
+///     }
+///   }
+///
+///   // there will be at most one terminal completion
+///   assert_eq!(completions.load(std::sync::atomic::Ordering::Relaxed), 1);
+/// }
+/// ```
+#[derive(Clone)]
+pub struct SharedRetryable<V> {
+    inner: Arc<tokio::sync::Mutex<Option<MaybeDone<V>>>>,
+}
+
+impl<V> Default for SharedRetryable<V> {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(tokio::sync::Mutex::new(None)),
+        }
+    }
+}
+
+/// Determine if an error is transient or permanent.
+pub trait Retryable {
+    fn is_permanent(&self) -> bool {
+        true
+    }
+}
+
+/// Retried task panicked, was cancelled, or never spawned (see [`SharedRetryable::try_restart`]).
+#[derive(Debug, PartialEq, Eq)]
+pub struct RetriedTaskPanicked;
+
+impl<T, E1> SharedRetryable<Result<T, E1>>
+where
+    T: Clone + std::fmt::Debug + Send + 'static,
+    E1: Clone + Retryable + std::fmt::Debug + Send + 'static,
+{
+    /// Restart a previously failed operation unless it already completed with a terminal result.
+    ///
+    /// Many futures can call this function and and get the terminal result from an earlier attempt
+    /// or start a new attempt, or join an existing one.
+    ///
+    /// Compared to `Self::try_restart`, this method also spawns the future to run, which would
+    /// otherwise have to be done manually.
+    pub async fn try_restart_spawn<F, Fut, E2>(&self, retry_with: F) -> Result<T, E2>
```

Comment:
Sadly these methods cannot be made inaccessible with `#[cfg(doctest)]` because for doc tests a dev build (without `-C test`) is used? I'll move the doc comment as a test case then.

Reply:
However if I do this, I cannot include the "stress test" example, except as a test case; I'll add it in a separate module not to grow this one too long.

Reply:
Well, what do you think, is a stress test example useful? I think the current impl is quite straightforward, everything is behind that one mutex.

Reply:
Probably not needed.